Level 2 Materials.

WilkAndy opened the pull request:
Refs #6845

The Material architecture in the Porous Flow module is as follows.

(1) Level 1 Materials.  A user typically chooses one of these for their
particular simulation type, depending upon the overall type
of simulation.  For instance, there are Materials suitable for:
 (a) 1-phase saturated+unsaturated flow with porepressure as the primary nonlinear variable
 (b) 1-phase saturated+unsaturated flow with log(mass density) as the primary nonlinear variable
 (c) 2-phase flow with the two porepressures being the primary nonlinear variables
 (d) 2-phase flow with (porepressure0, saturation1) being the primary nonlinear variables
The purpose of these Level 1 Materials is to calculate, for each phase,
the porepressure, the saturation, gradients of these, and derivatives of
these wrt to the primary variables.  Also temperature, gradients of
temperature and derivatives of these wrt the primary variables.  Also
the mass fraction matrix, gradients and derivatives.  All these things
are calculated at the quadpoints and potentially the nodes depending on
whether we'll need nodal quantities in the mass-lumping and full-upwinding
in the Kernels.

(2) Level 2 Materials.  A user typically chooses many of these.
They are designed to get information from the Level 1 Materials and
calculate equation-of-state related things, such as fluid-phase density,
viscosity, relative permeability, and also solid-phase porosity.
They are coded so as to be independent of the Level 1 description, eg
a user can use the water density Material in both the 1-phase and
the 2-phase situation.  Typically these Level 2 Materials compute derivatives
of density, viscosity, etc, with respect to porepressures, saturations,
etc, independently of what the actual primary nonlinear variables are.

(3) Level 3 Materials.  These join the Level 2 Materials into std::vectors
for use in Kernels, and also join the derivative information of Level 2 Materials
into std::vectors that contain derivatives wrt the primary nonlinear variables

WilkAndy committed:
Level 2 Materials.

Refs #6845

The Material architecture in the Porous Flow module is as follows.

(1) Level 1 Materials.  A user typically chooses one of these for their
particular simulation type, depending upon the overall type
of simulation.  For instance, there are Materials suitable for:
 (a) 1-phase saturated+unsaturated flow with porepressure as the primary nonlinear variable
 (b) 1-phase saturated+unsaturated flow with log(mass density) as the primary nonlinear variable
 (c) 2-phase flow with the two porepressures being the primary nonlinear variables
 (d) 2-phase flow with (porepressure0, saturation1) being the primary nonlinear variables
The purpose of these Level 1 Materials is to calculate, for each phase,
the porepressure, the saturation, gradients of these, and derivatives of
these wrt to the primary variables.  Also temperature, gradients of
temperature and derivatives of these wrt the primary variables.  Also
the mass fraction matrix, gradients and derivatives.  All these things
are calculated at the quadpoints and potentially the nodes depending on
whether we'll need nodal quantities in the mass-lumping and full-upwinding
in the Kernels.

(2) Level 2 Materials.  A user typically chooses many of these.
They are designed to get information from the Level 1 Materials and
calculate equation-of-state related things, such as fluid-phase density,
viscosity, relative permeability, and also solid-phase porosity.
They are coded so as to be independent of the Level 1 description, eg
a user can use the water density Material in both the 1-phase and
the 2-phase situation.  Typically these Level 2 Materials compute derivatives
of density, viscosity, etc, with respect to porepressures, saturations,
etc, independently of what the actual primary nonlinear variables are.

(3) Level 3 Materials.  These join the Level 2 Materials into std::vectors
for use in Kernels, and also join the derivative information of Level 2 Materials
into std::vectors that contain derivatives wrt the primary nonlinear variables

WilkAndy commented:
yep, this is the big one!   Lots of nasty equations-of-state.

dschwen commented on modules/porous_flow/include/materials/PorousFlowDensityConstBulk.h:
```diff
@@ -0,0 +1,71 @@
+   * @param pressure gas pressure (Pa)
+   * @return derivative of density (kg/m^3) with respect to pressure
+   */
+  Real dDensity_dP(Real pressure) const;
```
For the record, I'm not a fan of these method names with underscores. But I don't want to be the `PorousFlowMethodNameDictator` :-)

dschwen commented on modules/porous_flow/include/materials/PorousFlowSimpleCO2.h:
```diff
@@ -0,0 +1,261 @@
+   *
+   * @return constants for Henry's constant (-)
+   */
+  std::vector<Real> henryConstants() const;
```
where will this be used (it is not called anywhere yet, is it?)

WilkAndy replied:
Henry's constant dictate how much of one species (CO2) is dissolved in a particular phase (water), and as such they affect the density and viscosity of the phase.  At present i think we haven't any tests, kernels or materials that actually use these things, but in the near future (within a month i hope) we will be using them a lot.

dschwen commented on modules/porous_flow/include/materials/PorousFlowSimpleCO2.h:
```diff
@@ -0,0 +1,261 @@
+  const Real _p_critical;
+
+  /// Critical temperature (C)
+  const Real _t_critical;
```
usually we name a Temperature `T` and a time `t`

dschwen commented on modules/porous_flow/src/materials/PorousFlowBrine.C:
```diff
@@ -0,0 +1,206 @@
+  Real Mbrine = 1.0 / (xnacl / _Mnacl + (1.0 - xnacl) / _Mh2o);
+  Real Xnacl = xnacl / _Mnacl * Mbrine;
+
+  n11 = -54.2958 - 45.7623 * std::exp(-9.44785e-4 * pbar);
```
just a suggestion, but maybe a literature reference would be helpful if somebody ever wants to trace the origin of all these numbers.

WilkAndy replied:
I put all the references that i had into the .h and .C files.

WilkAndy replied:
I mean: "In the latest commit I put all the references that i had into the .h and .C files."

dschwen commented on modules/porous_flow/src/materials/PorousFlowCapillaryPressureLinear.C:
```diff
@@ -0,0 +1,37 @@
+void
+PorousFlowCapillaryPressureLinear::computeQpProperties()
+{
+  /// Capillary pressure and derivatives wrt phase saturation at the nodes
```
Wait... that requires a quadrature/order where the number of QPs is exactly the number of nodes, right? Are you checking this anywhere?

WilkAndy replied:
Number of QPs = Number of nodes.  YES, this is a major assumption of our code and is necessary because of the full-upwinding and mass lumping that we require.  I'm not sure how to check this.  Could you tell me: (1) how to check it; (2) where you think it best be checked (i suspect in the Level 1 Materials - but in their constructors, or somewhere else?)

dschwen replied:
The minimum in my opinion should be an assert on the sizes right here. The actual check is not straightforward, as you could have mixed element types (varying number of nodes). So you cannot do a simple check in a constructor. I think you may have to iterate over the mesh. Comments @permcody

permcody replied:
We already loop over all elements before the simulation to size certain
data structures. Look at ComputeMaxQps. We could probably store addition
pieces of data to make these types of checks more efficient during the
simulation.
On Mon, May 9, 2016 at 5:09 PM Daniel Schwen <notifications@github.com>
wrote:

> In modules/porous_flow/src/materials/PorousFlowCapillaryPressureLinear.C
> <https://github.com/idaholab/moose/pull/6889#discussion_r62588645>:
>
> > +  InputParameters params = validParams<PorousFlowCapillaryPressureBase>();
> > +  params.addRequiredParam<Real>("pc_max",  "Maximum capillary pressure (Pa)");
> > +  params.addClassDescription("This Material provides a linear capillary pressure");
> > +  return params;
> > +}
> > +
> > +PorousFlowCapillaryPressureLinear::PorousFlowCapillaryPressureLinear(const InputParameters & parameters) :
> > +  PorousFlowCapillaryPressureBase(parameters),
> > +  _pc_max(getParam<Real>("pc_max"))
> > +{
> > +}
> > +
> > +void
> > +PorousFlowCapillaryPressureLinear::computeQpProperties()
> > +{
> > +  /// Capillary pressure and derivatives wrt phase saturation at the nodes
>
> The minimum in my opinion should be an assert on the sizes right here. The
> actual check is not straightforward, as you could have mixed element types
> (varying number of nodes). So you cannot do a simple check in a
> constructor. I think you may have to iterate over the mesh. Comments
> @permcody <https://github.com/permcody>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6889/files/4f2d0ae39e837aeec61169a353951bd97d029281#r62588645>
>

WilkAndy replied:
In the Dictator (which every porous-flow simulation needs), could i do the loop over the mesh?  Not that i actually know how to do that, but it would be a good place.

for (elem = 0; elem < number_elements; ++elem)
  if (elem.numberQPs() != elem.numberNodes())
    CRASH

WilkAndy replied:
The Dictator is a GeneralUserObject, by the way, @permcody

permcody replied:
You could but we already do this loop in moose. We could store off extra
information along the way to make this check possible.
On Mon, May 9, 2016 at 5:48 PM Andy Wilkins <notifications@github.com>
wrote:

> In modules/porous_flow/src/materials/PorousFlowCapillaryPressureLinear.C
> <https://github.com/idaholab/moose/pull/6889#discussion_r62592519>:
>
> > +  InputParameters params = validParams<PorousFlowCapillaryPressureBase>();
> > +  params.addRequiredParam<Real>("pc_max",  "Maximum capillary pressure (Pa)");
> > +  params.addClassDescription("This Material provides a linear capillary pressure");
> > +  return params;
> > +}
> > +
> > +PorousFlowCapillaryPressureLinear::PorousFlowCapillaryPressureLinear(const InputParameters & parameters) :
> > +  PorousFlowCapillaryPressureBase(parameters),
> > +  _pc_max(getParam<Real>("pc_max"))
> > +{
> > +}
> > +
> > +void
> > +PorousFlowCapillaryPressureLinear::computeQpProperties()
> > +{
> > +  /// Capillary pressure and derivatives wrt phase saturation at the nodes
>
> The Dictator is a GeneralUserObject, by the way, @permcody
> <https://github.com/permcody>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6889/files/4f2d0ae39e837aeec61169a353951bd97d029281#r62592519>
>

WilkAndy replied:
So shall I submit an Issue which you will work on?

permcody replied:
Better put in an issue. I'm so busy I may not get to it right away. You
could if you want though :)
On Mon, May 9, 2016 at 5:52 PM Andy Wilkins <notifications@github.com>
wrote:

> In modules/porous_flow/src/materials/PorousFlowCapillaryPressureLinear.C
> <https://github.com/idaholab/moose/pull/6889#discussion_r62592977>:
>
> > +  InputParameters params = validParams<PorousFlowCapillaryPressureBase>();
> > +  params.addRequiredParam<Real>("pc_max",  "Maximum capillary pressure (Pa)");
> > +  params.addClassDescription("This Material provides a linear capillary pressure");
> > +  return params;
> > +}
> > +
> > +PorousFlowCapillaryPressureLinear::PorousFlowCapillaryPressureLinear(const InputParameters & parameters) :
> > +  PorousFlowCapillaryPressureBase(parameters),
> > +  _pc_max(getParam<Real>("pc_max"))
> > +{
> > +}
> > +
> > +void
> > +PorousFlowCapillaryPressureLinear::computeQpProperties()
> > +{
> > +  /// Capillary pressure and derivatives wrt phase saturation at the nodes
>
> So shall I submit an Issue which you will work on?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6889/files/4f2d0ae39e837aeec61169a353951bd97d029281#r62592977>
>

WilkAndy replied:
OK.

dschwen commented on modules/porous_flow/src/materials/PorousFlowCapillaryPressureVGP.C:
```diff
@@ -0,0 +1,91 @@
+  if (pressure >= 0.0)
+    return 0.0;
+  else
+  {
```
that else is not really needed

dschwen commented on modules/porous_flow/src/materials/PorousFlowCapillaryPressureVGS.C:
```diff
@@ -0,0 +1,84 @@
+}
+
+PorousFlowCapillaryPressureVGS::PorousFlowCapillaryPressureVGS(const InputParameters & parameters) :
+  PorousFlowCapillaryPressureBase(parameters),
```
indent by 4 spaces

dschwen commented on modules/porous_flow/src/materials/PorousFlowCapillaryPressureVGS.C:
```diff
@@ -0,0 +1,84 @@
+Real
+PorousFlowCapillaryPressureVGS::effectiveSaturation(Real saturation) const
+{
+  return (saturation - _sat_lr)/(_sat_ls - _sat_lr);
```
operator spacing

dschwen commented on modules/porous_flow/src/materials/PorousFlowEffectiveFluidPressure.C:
```diff
@@ -0,0 +1,59 @@
+PorousFlowEffectiveFluidPressure::computeQpProperties()
+{
+  _pf_qp[_qp] = 0;
+  _pf_nodal[_qp] = 0;
```
I'd prefer `0.0` over `0` if those are floating point variables. Makes the code clearer IMO.

dschwen commented on modules/porous_flow/src/materials/PorousFlowMethane.C:
```diff
@@ -0,0 +1,122 @@
+Real
+PorousFlowMethane::viscosity(Real temperature) const
+{
+  Real a[6] = {2.968267e-1, 3.711201e-2, 1.218298e-5, -7.02426e-8, 7.543269e-11,
```
might as well const this

dschwen commented on modules/porous_flow/src/materials/PorousFlowPorosityConst.C:
```diff
@@ -0,0 +1,46 @@
+  _dporosity_nodal_dvar[_qp].resize(num_var, 0.0);
+  _dporosity_qp_dvar[_qp].resize(num_var, 0.0);
+  _dporosity_nodal_dgradvar[_qp].resize(num_var, RealGradient());
+  _dporosity_qp_dgradvar[_qp].resize(num_var, RealGradient());
```
using `assign` rather than `resize` would be a bit more expressive here (resize will only initialize the newly allocated entries, whereas assign makes it clear that you want the entire vector to be initialized with the second argument)

dschwen commented on modules/porous_flow/src/materials/PorousFlowPorosityHM.C:
```diff
@@ -0,0 +1,84 @@
+  _porosity_qp[_qp] = _biot + (_phi0 - _biot) * std::exp(-_vol_strain_qp[_qp] + _coeff * _pf_qp[_qp]);
+
+  _dporosity_qp_dvar[_qp].resize(_num_var, 0.0);
+  _dporosity_nodal_dvar[_qp].resize(_num_var, 0.0);
```
and here you can leave off the second argument. Initialization is not necessary, as you set each entry in the loop below

dschwen commented on modules/porous_flow/src/materials/PorousFlowPorosityHM.C:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  _dporosity_qp_dgradvar[_qp].resize(_num_var, RealGradient());
+  _dporosity_nodal_dgradvar[_qp].resize(_num_var, RealGradient());
```
same here. Yeah, this is picky. The generated code will be most likely identical. This calls should be no-ops for all but the first calls of this method. I'm just thinking of optimizing the expressiveness of the code.

dschwen commented on modules/porous_flow/src/materials/PorousFlowPorosityUnity.C:
```diff
@@ -0,0 +1,56 @@
+  _dporosity_nodal_dvar[_qp].resize(num_var, 0.0);
+  _dporosity_qp_dvar[_qp].resize(num_var, 0.0);
+  _dporosity_nodal_dgradvar[_qp].resize(num_var, RealGradient());
+  _dporosity_qp_dgradvar[_qp].resize(num_var, RealGradient());
```
same resize/assign suggestion

dschwen commented on modules/porous_flow/src/materials/PorousFlowRelativePermeabilityUnity.C:
```diff
@@ -0,0 +1,40 @@
+    _relative_permeability(declareProperty<Real>("PorousFlow_relative_permeability" + Moose::stringify(_phase_num))),
+    _drelative_permeability_ds(declarePropertyDerivative<Real>("PorousFlow_relative_permeability" + Moose::stringify(_phase_num), _saturation_variable_name))
+{
+  if (_phase_num >= _dictator_UO.numPhases())
```
this check seems to be made frequently. You might even add a method in the Dictator object (`checkPhaseNum(_phase_num)`) or have a material base class that owns the `_phase_num` and `PorousFlowDictator_UO` parameters and does the check. 

On that note I tried to simplify parameter names in phase field and _not_ have mixed case and long typenames in them. So this parameter would just be called `dictator ` in phase field. The fact that it is a user object name is already captured in the parameter type, and the fact that a `PorousFlowDictator` is requested should be in the doc string. But I guess that is a matter of taste.

dschwen commented on modules/porous_flow/src/materials/PorousFlowSimpleCO2.C:
```diff
@@ -0,0 +1,562 @@
+}
+
+PorousFlowSimpleCO2::PorousFlowSimpleCO2(const InputParameters & parameters) :
+    PorousFlowFluidPropertiesBase(parameters),
```
If `PorousFlowFluidPropertiesBase` inherited from the base class I suggested above you could also have a string member `_phase` that contains the stringified phase number. That would get rid of the twenty Moose::stringify calls.

dschwen commented on modules/porous_flow/src/materials/PorousFlowSimpleCO2.C:
```diff
@@ -0,0 +1,562 @@
+Real
+PorousFlowSimpleCO2::supercriticalViscosity(Real pressure, Real temperature) const
+{
+  Real b0[5] = {-1.958098980443E+01, 1.123243298270E+00, -2.320378874100E-02,
```
those should all be const. Also a literature reference as a comment might be helpful.

dschwen commented on modules/porous_flow/src/materials/PorousFlowWater.C:
```diff
@@ -0,0 +1,486 @@
+InputParameters validParams<PorousFlowWater>()
+{
+  InputParameters params = validParams<PorousFlowFluidPropertiesBase>();
+  params.addClassDescription("This Material calculates fluid properties for water (H20)");
```
Is this really _H twenty_ or should it be _H two O_?

dschwen commented on modules/porous_flow/src/materials/PorousFlowWater.C:
```diff
@@ -0,0 +1,486 @@
+Real
+PorousFlowWater::dDensity_dT(Real /*pressure*/, Real /*temperature*/) const
+{
+  return 0.; // TODO: not implemented yet
```
`0.0`

WilkAndy committed:
Implemented Daniel changes.

Except:
  did not change things like dDensity_dP (with underscores) as it looks good to me
  have not yet checked num QPS = num nodes (see issue #6940)
  have not made a base class that holds phase_num and does the stringify once only (opened a new issue #6938 to do this once everything is merged)
  have not changed _dictator_UO to _dictator (opened a new issue #6939 to do this for all PorousFlow classes, once everything is merged)

Refs #6845

permcody commented:
This looks OK. I'll merge this since I know you have much more coming. Please keep @dschwen's suggestions in mind for improving the code base.

WilkAndy commented:
Thanks @permcody , yep, i've opened a couple of issues as suggested by @dschwen which we'll address when the whole lot is merged (which will be soon, yippee!)
